[InstCombine] Add constant folding for hypot library calls

wentywenty opened the pull request:
This patch adds constant folding optimization for hypot(x, y) when both
arguments are constant values.

Implementation details:
- Only supports float and double precision, as std::hypot lacks support
  for extended precision types (fp128, x86_fp80, ppc_fp128, fp16)
- Respects errno semantics: folds when the function is marked as not
  accessing memory (intrinsic), fast-math is enabled, or when the result
  is finite / inputs are non-finite (hypot only sets errno on overflow
  with finite inputs)
- Follows the same pattern as other LibCall optimizations like remquo

The optimization handles:
- Basic cases: hypot(3.0, 4.0) -> 5.0
- Special values: Inf and NaN propagation
- Overflow: preserves calls in strict mode, folds in fast-math mode
- Strict FP mode: preserves calls to maintain FP exception semantics

Tests cover all major scenarios including precision handling, special
values, overflow behavior, and strictfp mode.

Related issue: https://github.com/llvm/llvm-project/issues/113711

wentywenty committed:
[InstCombine] Drop one-use check for (A & ~B) | (A ^ B) -> A ^ B

This patch removes the m_OneUse constraint for the fold:
  (A & ~B) | (A ^ B) -> A ^ B

Previously, this optimization was limited to cases where the operands had
a single use. However, replacing the 'or' with the existing 'xor' operand
is always profitable because it strictly removes one instruction (the 'or'),
regardless of whether the intermediate 'and' or 'xor' are used elsewhere.

Tests have been updated to reflect this multi-use optimization.

Signed-off-by: wentywenty <2321901849@qq.com>

nikic reviewed (changes requested):
Please see https://llvm.org/docs/InstCombineContributorGuide.html#precommit-tests.

All of your tests already fold without your patch: https://llvm.godbolt.org/z/ffs3MYGGT

wentywenty committed:
[InstCombine] Add constant folding for hypot library calls

This patch adds constant folding optimization for hypot(x, y) when both
arguments are constant values.

Implementation details:
- Only supports float and double precision, as std::hypot lacks support
  for extended precision types (fp128, x86_fp80, ppc_fp128, fp16)
- Respects errno semantics: folds when the function is marked as not
  accessing memory (intrinsic), fast-math is enabled, or when the result
  is finite / inputs are non-finite (hypot only sets errno on overflow
  with finite inputs)
- Follows the same pattern as other LibCall optimizations like remquo

The optimization handles:
- Basic cases: hypot(3.0, 4.0) -> 5.0
- Special values: Inf and NaN propagation
- Overflow: preserves calls in strict mode, folds in fast-math mode
- Strict FP mode: preserves calls to maintain FP exception semantics

Tests cover all major scenarios including precision handling, special
values, overflow behavior, and strictfp mode.

Signed-off-by: wentywenty <2321901849@qq.com>
